### DISCUSSION TITLE
[General] Precompile Calls

### DISCUSSION THREAD

**Discussion by 4NNNN** [zkSync-Community-Hub/zksync-developers#176] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

Testing precompiles and tried calling the sha precompile using the EfficientCall contract

### What is the specific issue or error you're encountering?

System Panics - I do know the VM panics if the system call isn't a **isSystem** flagged call tried **MimicCall** 'ing the precompile contract still couldn't test it out

### Can you share the error messages or logs you're receiving, if any?

Call(Normal) 0x36615cf349d7f6344891b1e7ca7c72883f5dc049           executeTransaction(bytes32, bytes32, tuple) Revert: Unknown revert reason  2439297
06:20:44 [INFO]     Call(Normal) 0x26b368c3ed16313ebd6660b72d8e4439a697cb0b                 0x544abfa0 Revert: Unknown revert reason  2399607
06:20:44 [INFO]       Call(Normal) 0x000000000000000000000000000000000000ffeb           0x  Error: Panic 0
06:20:44 [INFO]       Call(Normal) 0x000000000000000000000000000000000000ffe9           0x  Error: Panic 676
06:20:44 [INFO]       Call(Normal) 0x000000000000000000000000000000000000ffe8           0x  Error: Panic 128
06:20:44 [INFO]       Call(Normal) 0x000000000000000000000000000000000000ffe7           0x  Error: Panic 2360295
06:20:44 [INFO]       Call(Normal) 0x000000000000000000000000000000000000fff6           0x  Error: Panic 2
06:20:44 [INFO]   Call(Normal) System context

### Have you made any recent changes to the contract before encountering this issue?

- How can I call and test precompiles from a smart contract? 
- Should I mimic **Msg Value system contract** address and use the efficient call library (**staticcall** / **MimicCall** / or which function ) with the **isSystem** flagged true?
- Also referred to **SystemContractHelper.sol** - Why should i **packPrecompileParams** is there any specific reason behind it, If so how do I do it and is **precompileCall** the only way to call a precompile ?

### Are there any external libraries or contracts that your contract interacts with?

NA

### Can you provide the relevant portions of your contract code where the issue is occurring?

NA

### Have you tried to isolate the problem, and if so, what were the results?

Tried calling a custom precompile (0x19) as well as the sha precompile with the rust sdk using the era_call method
let response = call(precompile_address, inputs, data, &era_provider()).await?;
code:
```
pub async fn call(
    precompile_address: Address,
    _inputs: Option<&[&str]>,
    data: Option<Bytes>,
    provider: &Provider<Http>,
) -> Result<Bytes, ProviderError> {
    if let Some(data) = data {
        let request = Eip1559TransactionRequest::new()
            .to(precompile_address)
            .data(data);
        let transaction: TypedTransaction = request.into();
        let encoded_output = Middleware::call(&provider, &transaction, None).await?;
        Ok(encoded_output)
    } else {
        panic!("inputs or data must be provided")
    }
}
```

It works!

### What steps have you already taken to try to resolve the issue?

06:26:22 [INFO] Call(Normal) Account Code Storage                                 getCodeHash(uint256)   4227857361
06:26:22 [INFO]   Call(Normal) System context                                             0x29f172ad   4227854148
06:26:22 [INFO]   Call(Normal) System context                                             0x06bed036   4227844005
06:26:22 [INFO]     Call(Normal) Keccak                                               00000000   4161782016
06:26:22 [INFO]   Call(Normal) Account Code Storage                                 getRawCodeHash(address)   78749370
06:26:22 [INFO]   Call(Normal) System context                                       setTxOrigin(address)   78748425
06:26:22 [INFO]   Call(Normal) Known code storage                                   markFactoryDeps(bool, bytes32[])   78743133
06:26:24 [INFO]   Call(Mimic) Msg Value System Contract                                  0x27f19277   78741621
06:26:24 [INFO]     Call(Mimic) 0x0000000000000000000000000000000000000019                 0x27f19277   77507892

"success"

**Comment by 4NNNN** [comment]


**Reply by Romsters** [reply, marked as answer]
@4NNNN  You can call supported precompiles like this:
```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

contract GetHash {
    function getHash(bytes memory data) public pure returns (bytes32) {
        return sha256(data);
    }
}
```

Ethereum cryptographic primitives like ecrecover, keccak256, and sha256 are supported as precompiles. No actions are required from your side as all the calls to the precompiles are done by the compilers under the hood.

**Comment by patelkaren** [comment]
To call and test precompiles from a smart contract, you can use the efficient call library with the isSystem flag set to true. This flag indicates that the call is a system call, which is necessary for interacting with precompiled contracts.

Regarding the SystemContractHelper.sol, the packPrecompileParams function is used to encode the parameters required by the precompile into the correct format for the call. It ensures that the input data is properly formatted before being passed to the precompile.

Using the precompileCall function is indeed one way to call a precompile from a smart contract, but it's not the only way. You can also use regular contract calls with the efficient call library, as long as the isSystem flag is set appropriately.

In summary, to call and test precompiles from a smart contract:

Use the efficient call library with the isSystem flag set to true.
Ensure that the input parameters are properly encoded using packPrecompileParams if necessary.
Use precompileCall or regular contract calls with the efficient call library to interact with the precompile.
By following these steps, you should be able to call and test precompiles effectively from your smart contract.

**Reply by 4NNNN** [reply]
@patelkaren A custom precompile function call can only be done like a normal contract call after deployment. If you want the system contracts to do a static call / mimic call, you just have to initiate a in-memory node with the custom precompile contract address configured. Honestly, the precompile params packing is a lot of work in order to just test I would rather deploy a seperate contract and benchmark the precompile! Anyways this issue was solved months ago, thanks tho!

**Comment by kaaron9** [comment]
To call and test precompiles from a smart contract, you can use the efficient call library and mimic the system contract address with the isSystem flagged set to true. This allows you to interact with precompiles effectively. Additionally, you may need to pack precompile parameters using packPrecompileParams. PrecompileCall is one way to call a precompile, but ensure that you're using the correct method for your specific use case.
